### PR TITLE
Expose extension methods to wrap and unwrap adapters

### PR DESCRIPTION
**Pull request by twsouthwick**
As it currently stands, we only support using casting to convert between the two representations. This change exposes the extension methods we are using internally to expose the ASP.NET Core or System.Web objects.

Fixes https://github.com/dotnet/aspnetcore/issues/51119

### CODE REVIEW

**Comment by twsouthwick** [comment]
@davidfowl Here are the extension methods. I'm not happy with the name, though, as it feels fairly generic (but maybe sufficient?).

File: src/Microsoft.AspNetCore.SystemWebAdapters/SystemWebAdapterExtensions.cs [hunk 1]
```diff
@@ -2,47 +2,38 @@
...
-using System.Collections.Generic;
 using System.Diagnostics.CodeAnalysis;
 using System.Web;
 using Microsoft.AspNetCore.Http.Features;
-using Microsoft.AspNetCore.SystemWebAdapters.Internal;
 
 namespace Microsoft.AspNetCore.SystemWebAdapters;
 
-internal static class InternalSystemWebExtensions
+public static class SystemWebAdapterExtensions
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SystemWebAdapters/SystemWebAdapterExtensions.cs:11]: Do we make the method names more specific? Seeing GetAdapter looks a little hard to understand. 

-  `GetHttpRequestAdapter()`
- `GetSystemWebRequest()`

Feels API review worthy.

Comment (twsouthwick) [reply src/Microsoft.AspNetCore.SystemWebAdapters/SystemWebAdapterExtensions.cs]: reviewed and updated

File: src/Microsoft.AspNetCore.SystemWebAdapters/SystemWebAdaptersConversionExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    public static HttpResponse AsSystemWeb(this HttpResponseCore response)
+    {
+        ArgumentNullException.ThrowIfNull(response);
+
+        return response.HttpContext.AsSystemWeb().Response;
+    }
+
+    public static HttpResponseCore AsAspNetCore(this HttpResponse response)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.SystemWebAdapters/SystemWebAdaptersConversionExtensions.cs:25]: Sort these either by framework or by type (request/response/context)

File: src/Microsoft.AspNetCore.SystemWebAdapters/SystemWebAdaptersConversionExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        ArgumentNullException.ThrowIfNull(response);
+
+        return response;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.SystemWebAdapters/SystemWebAdaptersConversionExtensions.cs:29]: This relies on the implicit converter? But going the other way the implicit converters rely on the extensions? Be consistent.

File: src/Microsoft.AspNetCore.SystemWebAdapters/SystemWebAdaptersConversionExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    public static HttpContextCore AsAspNetCore(this HttpContext context)
+    {
+        ArgumentNullException.ThrowIfNull(context);
+
+        return context;
+    }
+
+    public static HttpRequestCore GetAspNetCoreRequest(this HttpRequest request)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.SystemWebAdapters/SystemWebAdaptersConversionExtensions.cs:54]:
```suggestion
    public static HttpRequestCore AsAspNetCore(this HttpRequest request)
```

**Review by Tratcher — CHANGES_REQUESTED** [review]

File: src/Microsoft.AspNetCore.SystemWebAdapters/SystemWebAdaptersConversionExtensions.cs (hunk 2 above)
Comment (twsouthwick) [reply src/Microsoft.AspNetCore.SystemWebAdapters/SystemWebAdaptersConversionExtensions.cs]: done

File: src/Microsoft.AspNetCore.SystemWebAdapters/SystemWebAdaptersConversionExtensions.cs (hunk 4 above)
Comment (twsouthwick) [reply src/Microsoft.AspNetCore.SystemWebAdapters/SystemWebAdaptersConversionExtensions.cs]: thanks - must have missed that one

File: src/Microsoft.AspNetCore.SystemWebAdapters/SystemWebAdaptersConversionExtensions.cs (hunk 3 above)
Comment (twsouthwick) [reply src/Microsoft.AspNetCore.SystemWebAdapters/SystemWebAdaptersConversionExtensions.cs]: It needs a private field to do the conversion, so I've made the implicit operators the source of truth.

Comment (twsouthwick) [reply src/Microsoft.AspNetCore.SystemWebAdapters/SystemWebAdaptersConversionExtensions.cs]: switched to an internal property

File: src/Microsoft.AspNetCore.SystemWebAdapters/HttpContextBase.cs [hunk 5]
```diff
@@ -57,7 +57,7 @@ public virtual IPrincipal User
         public virtual object? GetService(Type serviceType) => throw new NotImplementedException();
 
         [return: NotNullIfNotNull(nameof(context))]
-        public static implicit operator HttpContextBase?(HttpContextCore? context) => context?.GetAdapterBase();
+        public static implicit operator HttpContextBase?(HttpContextCore? context) => context?.AsSystemWebBase();
```
[the remark this answers is not in the document]
Comment (twsouthwick) [reply src/Microsoft.AspNetCore.SystemWebAdapters/HttpContextBase.cs]: There's no way to get a HttpContext from HttpContextBase. If we were to do this, we'd have to rework everything on top of HttpContextBase

File: src/Microsoft.AspNetCore.SystemWebAdapters/HttpRequest.cs [hunk 6]
```diff
@@ -290,7 +290,7 @@ private static void WriteTo(Stream source, Stream destination)
         public void Abort() => _request.HttpContext.Abort();
 
         [return: NotNullIfNotNull(nameof(request))]
-        public static implicit operator HttpRequest?(HttpRequestCore? request) => request.GetAdapter();
+        public static implicit operator HttpRequest?(HttpRequestCore? request) => request?.HttpContext.AsSystemWeb().Request;
 
         [return: NotNullIfNotNull(nameof(request))]
         public static implicit operator HttpRequestCore?(HttpRequest? request) => request?._request;
```
[the remark this answers is not in the document]
Comment (twsouthwick) [reply src/Microsoft.AspNetCore.SystemWebAdapters/HttpRequest.cs]: I've swapped this per our discussion

File: src/Microsoft.AspNetCore.SystemWebAdapters/FeatureCollectionExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using Microsoft.AspNetCore.Http.Features;
+
+namespace Microsoft.AspNetCore.SystemWebAdapters;
+
+internal static class FeatureCollectionExtensions
+{
+    internal static TFeature GetRequired<TFeature>(this IFeatureCollection features)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.SystemWebAdapters/FeatureCollectionExtensions.cs:11]: Is GetRequiredFeature not available?
https://github.com/dotnet/aspnetcore/blob/64f95ef2377403e075cced875363419755d317a4/src/Extensions/Features/src/FeatureCollectionExtensions.cs#L21

**Review by Tratcher — APPROVED** [review]

File: src/Microsoft.AspNetCore.SystemWebAdapters/FeatureCollectionExtensions.cs (hunk 7 above)
Comment (twsouthwick) [reply src/Microsoft.AspNetCore.SystemWebAdapters/FeatureCollectionExtensions.cs:11]: We're targeting .NET 6 where it is not. I've looked into switching to that on .NET 7+, but we have internals visible to and it ends up causing compiler warnings of multiple resolved types.

**Pull request merged**